ios/development-rules.md

```diff
@@ -0,0 +1,426 @@
+    }
+
+    public func execute(username: String, password: String) async throws {
+        try await someOtherUseCase.execute()
+        try await authRepository.login(username: username, password: password)
+    }
+}
+```
+- You can then register your use case in the application layer in `DependencyInjection/UseCase+Injection.swift`
+```
+public extension Resolver {
+    static func registerUseCases() {
+
+        // ...
+
+        register { LoginUseCaseImpl(authRepository: resolve(), someOtherUseCase: resolve()) as LoginUseCase }
+
+        // ...
+    }
+}
+```
+
+#### Repositories
+- We define repositories under the `Repositories` directory in each logical part
+- The domain layer defines protocols for these repositories – functions and their interfaces
+- You should use CRUD naming for repository functions that handle objects
+- Repository example:
+```
+public protocol AuthRepository {
+    func login(username: String, password: String) async throws
+    func logout() async throws
+    func register(username: String, password: String) async throws
+}
+```
+- The data layer provides repository implementations, as those are dependent on more data-level stuff
+
+### Data Layer
+- Provides required data via Repositories and Providers from database / network / etc.
+- Our data layer is separated into `Toolkits` and `Providers`
+- There is a toolkit for every logical part of the domain layer defining a repository protocol
+- These toolkits provide repository implementations and relevant `NetworkModels`, `DatabaseModels`, `NetworkEndpoints`, etc.
+- `Providers` define provider protocols and implementations
+- You should also stick with CRUD naming in the entire data layer
+
+#### Providers and their implementations
+- We use providers for handling each data source
+- A provider is defined as a protocol
+- Each provider can have multiple implementations, for example, one using a third-party library, and another one using our own custom implementation
+- Dependency Injection is again responsible for selecting the correct implementation
+- Provider example:
+```
+public protocol DatabaseProvider {
+    func read<T>(_ type: T.Type, id: Int) throws -> T
+    func update<T>(_ object: T, model: UpdateModel) throws –> T
+    func delete<T>(_ object: T) throws
+    func deleteAll() throws
+}
+```
+- Provider implementation example:
+```
+public struct RealmDatabaseProvider {
+
+    // ... some setup and context ...
+}
+
+extension RealmDatabaseProvider: DatabaseProvider {
+    public func read<T>(_ type: T.Type, id: Int) throws -> T {
+        // ...
+        if let object = realm.object(ofType: realmType.self, forPrimaryKey: id) as? T {
+            return object
+        } else {
+            throw SomeError
+        }
+    }
+
+    // ...
+}
+```
+- Just like with use cases, you register the provider implementation in the application layer in `DependencyInjection/Providers+Injection.swift`
+
+#### Repository implementations
```

Comment:
I would move the repository section above providers section, as we are first using the repository and then some provider